cns/restserver/ipam.go

```diff
@@ -478,101 +634,194 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+	podIPInfo := make([]cns.PodIpInfo, len(desiredIPAddresses))
 	service.Lock()
 	defer service.Unlock()
-
-	for _, ipConfig := range service.PodIPConfigState {
-		if ipConfig.IPAddress == desiredIPAddress {
-			switch ipConfig.GetState() { //nolint:exhaustive // ignoring PendingRelease case intentionally
-			case types.Assigned:
-				// This IP has already been assigned, if it is assigned to same pod, then return the same
-				// IPconfiguration
-				if ipConfig.PodInfo.Key() == podInfo.Key() {
-					logger.Printf("[AssignDesiredIPConfig]: IP Config [%+v] is already assigned to this Pod [%+v]", ipConfig, podInfo)
-				} else {
-					return podIpInfo, errors.Errorf("[AssignDesiredIPConfig] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
-				}
-			case types.Available, types.PendingProgramming:
-				// This race can happen during restart, where CNS state is lost and thus we have lost the NC programmed version
-				// As part of reconcile, we mark IPs as Assigned which are already assigned to Pods (listed from APIServer)
-				if err := service.assignIPConfig(ipConfig, podInfo); err != nil {
-					return podIpInfo, err
+	// creating a map for the loop to check to see if the IP in the pool is one of the desired IPs
+	IPMap := make(map[string]string)
+	for _, desiredIP := range desiredIPAddresses {
+		IPMap[desiredIP] = desiredIP
+	}
+	ncMap := make(map[string]cns.IPConfigurationStatus)
+	assignedMap := make(map[string]cns.IPConfigurationStatus)
+
+	for _, ipConfig := range service.PodIPConfigState { //nolint:gocritic // ignore copy
+		_, found := IPMap[ipConfig.IPAddress]
+		// keep searching until the all desired IPs are found
+		if !found {
+			continue
+		}
+		switch ipConfig.GetState() { //nolint:exhaustive // ignoring PendingRelease case intentionally
+		case types.Assigned:
+			// This IP has already been assigned, if it is assigned to same pod, then return the same
+			// IPconfiguration
+			if ipConfig.PodInfo.Key() == podInfo.Key() {
+				logger.Printf("[AssignDesiredIPConfigs]: IP Config [%+v] is already assigned to this Pod [%+v]", ipConfig, podInfo)
+				assignedMap[ipConfig.NCID] = ipConfig
+				err := service.populateIPConfigInfoUntransacted(ipConfig, &podIPInfo[len(assignedMap)])
+				if len(assignedMap) == len(desiredIPAddresses) {
+					return podIPInfo, err
 				}
-			default:
-				return podIpInfo, errors.Errorf("[AllocateDesiredIPConfig] Desired IP is not available %+v", ipConfig)
+			} else {
+				logger.Errorf("[AssignDesiredIPConfigs] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
+				//nolint:goerr113 // return error
+				return podIPInfo, fmt.Errorf("IP already assigned to another pod")
+			}
+		case types.Available, types.PendingProgramming:
+			// This race can happen during restart, where CNS state is lost and thus we have lost the NC programmed version
```

Comment:
This is getting called outside of this for loop in the next for loop. To avoid assigning IPs and then needing to release them again I just add them to a map here (now changed to a slice) to be looped through and added